src/factor_nodes/poisson.jl

```diff
@@ -0,0 +1,69 @@
+export Poisson
+
+"""
+Description:
+    Poisson factor node
+
+    Real scalars
+    l > 0 (rate)
+
+    f(out, l) = Poisson(out|l) = 1/(x!) * l^x * exp(-l)
+
+Interfaces:
+    1. out
+    2. l
+
+Construction:
+    Poisson(id=:some_id)
+"""
+mutable struct Poisson <: SoftFactor
+    id::Symbol
+    interfaces::Vector{Interface}
+    i::Dict{Symbol,Interface}
+
+    function Poisson(out, l, id=generateId(Poisson))
+        @ensureVariables(out, l)
+        self = new(id, Array{Interface}(undef, 2), Dict{Symbol, Interface}())
+        addNode!(currentGraph(), self)
+        self.i[:out] = self.interfaces[1] = associate!(Interface(self), out)
+        self.i[:l] = self.interfaces[2] = associate!(Interface(self), l)
+        return self
+    end
+end
+
+ForneyLab.slug(::Type{Poisson}) = "Poisson"
+
+ForneyLab.format(dist::ProbabilityDistribution{Univariate, Poisson}) = "$(slug(Poisson))(l=$(format(dist.params[:l])))"
+
+ForneyLab.ProbabilityDistribution(::Type{Univariate}, ::Type{Poisson}; l=1.0) = ProbabilityDistribution{Univariate, Poisson}(Dict(:l=>l))
+ForneyLab.ProbabilityDistribution(::Type{Poisson}; l=1.0) = ProbabilityDistribution{Univariate, Poisson}(Dict(:l=>l))
+
+ForneyLab.dims(dist::ProbabilityDistribution{Univariate, Poisson}) = 1
+
+ForneyLab.vague(::Type{Poisson}) = ProbabilityDistribution(Univariate, Poisson, l=huge)
+
+ForneyLab.isProper(dist::ProbabilityDistribution{Univariate, Poisson}) = (0 < dist.params[:l] < huge)
+
+ForneyLab.unsafeMean(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+ForneyLab.unsafeVar(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+ForneyLab.unsafeLogMean(dist::ProbabilityDistribution{Univariate, Poisson}) = dist.params[:l]
+
+# TODO: skip the implementation of entropy and average energy for now, they require additional work
+
+# Entropy functional
+function differentialEntropy(dist::ProbabilityDistribution{Univariate, Poisson})
+    l = clamp(dist.params[:l], tiny, huge)
+    h(l, j=20) = sum([(l)^(k)*log(factorial(k))/factorial(k) for k in collect(0:j)])
+    l*(1-log(l)) + exp(-l)*h(l)
+end
+
+# Average energy functional
+function ForneyLab.averageEnergy(::Type{Poisson}, marg_out::ProbabilityDistribution{Univariate}, marg_l::ProbabilityDistribution{Univariate})
+    # approximation of expectation of logX!
+    h(j) = sum([binomial(j, k)*(-1)^(-k)*log(factorial(k)) for k in collect(0:j)])
+    Elog_fact(λ, lim=20) = sum([(-λ)^j/factorial(j)*h(j) for j in collect(0:lim)])
```

Comment:
Please move the definitions for the `Elog_fact()` and `h()` functions to the outer scope; there is no need to define it within the `averageEnergy()` function. Also, a more descriptive naming and specific argument typing would be appreciated.